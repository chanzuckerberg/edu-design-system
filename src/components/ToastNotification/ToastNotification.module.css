@import '../../design-tokens/mixins.css';

/*------------------------------------*\
    # TOAST
\*------------------------------------*/

.toast {
  width: 100%;
  max-width: 24rem;
  display: flex;
  gap: calc(var(--eds-spacing-size-2) * 1px);
  padding: calc(var(--eds-spacing-size-2) * 1px);
  align-items: top;

  border: calc(var(--eds-spacing-size-quarter) * 1px) solid;
  border-left: calc(var(--eds-spacing-size-quarter) * 1px) solid;
  border-radius: calc(var(--eds-theme-border-radius-objects-sm) * 1px);


  &.toast--status-critical {
    color: var(--eds-theme-color-text-utility-critical);
    background-color: var(--eds-theme-color-background-utility-critical-low-emphasis);
  }

  &.toast--status-favorable {
    color: var(--eds-theme-color-text-utility-favorable);
    background-color: var(--eds-theme-color-background-utility-favorable-low-emphasis);
  }
}

.toast__icon {
  flex-shrink: 0;
}

.toast__body {
  flex-grow: 2;
}

.toast__dismiss-button {
  flex-shrink: 0;
  margin-top: calc(var(--eds-spacing-size-1) * -1px);
  margin-right: calc(var(--eds-spacing-size-1) * -1px);
}