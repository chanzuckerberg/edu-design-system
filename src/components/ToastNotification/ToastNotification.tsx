import clsx from 'clsx';
import React, { useEffect } from 'react';

import getIconNameFromStatus from '../../util/getIconNameFromStatus';
import type { Status } from '../../util/variant-types';
import Button from '../Button';
import Icon from '../Icon';
import Text from '../Text';

import styles from './ToastNotification.module.css';

export type ToastNotificationProps = {
  // Component API
  /**
   * Additional class names that can be appended to the component, passed in for styling.
   */
  className?: string;
  /**
   * Callback when notification is dismissed. When passed in, renders banner with a close icon in the top right.
   */
  onDismiss?: () => void;
  /**
   * Length of time to wait until `onDismiss` is called
   */
  timeout?: number;
  // Design API
  /**
   * Determines whether the toast notification will dismiss on its own, or due to user action. When set to `"auto"`,
   * it will dismiss after 8 seconds.
   *
   * **Default is `"manual"`**.
   */
  dissmissType?: 'manual' | 'auto';
  /**
   * Keyword to characterize the state of the notification
   */
  status?: Extract<Status, 'favorable' | 'critical'>;
  /**
   * The title/heading of the notification
   */
  title: string;
};

/**
 * `import {ToastNotification} from "@chanzuckerberg/eds";`
 *
 * Toasts display brief, temporary notifications. They're meant to be noticed without disrupting a user's experience or requiring an action to be taken.
 */
export const ToastNotification = ({
  className,
  dissmissType = 'manual',
  onDismiss,
  status = 'favorable',
  timeout = 8000,
  title,
  ...other
}: ToastNotificationProps) => {
  const componentClassName = clsx(
    styles['toast'],
    status && styles[`toast--status-${status}`],
    className,
  );

  useEffect(() => {
    const expireId =
      dissmissType === 'auto'
        ? setTimeout(() => {
            onDismiss && onDismiss();
          }, timeout)
        : undefined;

    return () => clearTimeout(expireId);
  }, [onDismiss, dissmissType, timeout]);

  return (
    <div className={componentClassName} {...other}>
      <Icon
        className={styles['toast__icon']}
        name={getIconNameFromStatus(status)}
        purpose="decorative"
        size="1.875rem"
      />
      <div className={styles['toast__body']}>
        <Text as="span" className={styles['toast__text']} preset="title-md">
          {title}
        </Text>
      </div>
      {onDismiss && (
        <Button
          aria-label="close"
          className={styles['toast__dismiss-button']}
          context="default"
          icon="close"
          iconLayout="icon-only"
          onClick={onDismiss}
          rank="tertiary"
          variant="neutral"
        >
          Close
        </Button>
      )}
    </div>
  );
};
