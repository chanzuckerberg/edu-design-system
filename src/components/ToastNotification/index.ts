export { ToastNotification as default } from './ToastNotification';
