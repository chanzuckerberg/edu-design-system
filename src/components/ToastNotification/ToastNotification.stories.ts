import type { StoryObj, Meta } from '@storybook/react';
import type { ComponentProps } from 'react';

import { ToastNotification } from './ToastNotification';

export default {
  title: 'Components/V2/ToastNotification',
  component: ToastNotification,
  parameters: {
    badges: ['intro-1.0', 'current-2.0'],
  },
  argTypes: { onDismiss: { action: 'dismissed' } },
  args: {
    title: "You've got a temporary notification!",
  },
} as Meta<Args>;

type Args = ComponentProps<typeof ToastNotification>;

export const Default: StoryObj<Args> = {};

export const Favorable: StoryObj<Args> = {
  args: {
    status: 'favorable',
  },
};

export const Critical: StoryObj<Args> = {
  args: {
    status: 'critical',
  },
};

export const NotDismissable: StoryObj<Args> = {
  args: {
    ...Default.args,
    onDismiss: undefined,
  },
};
