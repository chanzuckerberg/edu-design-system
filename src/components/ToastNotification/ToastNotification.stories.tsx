import { Transition } from '@headlessui/react';
import type { StoryObj, Meta } from '@storybook/react';
import React from 'react';

import type { ComponentProps } from 'react';

import { ToastNotification } from './ToastNotification';

import Button from '../Button';

export default {
  title: 'Components/ToastNotification',
  component: ToastNotification,
  parameters: {
    layout: 'centered',
    badges: ['api-2.0', 'theme-2.0'],
  },
  argTypes: {
    onDismiss: { action: 'trigger dismiss' },
    timeout: { table: { disable: true } },
  },
  args: {
    title: "You've got a temporary notification!",
    className: 'w-96',
  },
} as Meta<Args>;

type Args = ComponentProps<typeof ToastNotification>;
type Story = StoryObj<Args>;

export const Default: Story = {};

export const Favorable: Story = {
  args: {
    status: 'favorable',
  },
};

/**
 * Notifications can have different status, to indicate errors or destructive actions have completed.
 */
export const Critical: Story = {
  args: {
    status: 'critical',
  },
};

/**
 * We can restrict the ability to dismiss the notification by not specifying the `onDismiss` method.
 */
export const NotDismissable: Story = {
  args: {
    ...Default.args,
    onDismiss: undefined,
  },
};

/**
 * Tooltips can be instructed to auto-close after a certain period. After the timeout, the component will call the defined
 * `onDismiss` method. The behavior of the dissmisal is left up to the user, which allows for complete control.
 */
export const AutoDismiss: Story = {
  args: {
    ...Default.args,
    dismissType: 'auto',
    timeout: 500,
    onDismiss: () => console.log('trigger onDismiss'),
  },
};

let toastId = 0;
const ToastNotificationManager = (args: Args) => {
  const [toasts, setToasts] = React.useState<
    { id: number | string; text: string; show?: boolean }[]
  >([]);

  // TODO: clean up `toasts` after .show is set to false (using useEffect? and .debounce)
  // - In a production implementation, you can filter out any toasts where show=false

  return (
    <div className="flex h-[90vh] w-[90vw] items-center justify-center">
      <Button
        onClick={() => {
          setToasts([
            ...toasts,
            {
              id: toastId++,
              text: 'New Toast',
              show: true,
            },
          ]);
        }}
      >
        Trigger A Toast Notification
      </Button>
      <div
        className="absolute bottom-0 left-0 flex flex-col gap-spacing-size-2"
        id="toast-container"
      >
        {toasts.map((toast) => (
          <Transition
            appear
            as="div"
            enter="transition-all duration-medium"
            enterFrom="opacity-0 transform-gpu scale-0"
            enterTo="opacity-100 transform-gpu scale-100"
            key={toast.id}
            leave="ease-in-out transition-all duration-medium"
            leaveFrom="opacity-100 transform-gpu translate-x-[0px]"
            leaveTo="opacity-0 transform-gpu translate-x-[-100%]"
            show={toast.show}
          >
            <ToastNotification
              {...args}
              dismissType="auto"
              onDismiss={() => {
                setToasts(
                  toasts.map((thisToast) => {
                    return thisToast.id === toast.id
                      ? { ...thisToast, show: false }
                      : thisToast;
                  }),
                );
              }}
              title={'You got a new toast: ' + toast.text + toast.id}
            />
          </Transition>
        ))}
      </div>
    </div>
  );
};

/**
 * This implementation example shows how you can use toasts with state to handle multiple, stacking notifications.
 *
 * For a full, production-ready implementation, clean up any toasts with show=false after the animation has completed.
 * - Consider using lodash.debounce to time the re-render, and useEffect that watches the list of toasts
 * - Any debouncing should map to whatever duration is used in `Transition`
 *
 * Here, we use `<Transition>` provided by [HeadlessUI](https://github.com/chanzuckerberg/edu-design-system/blob/main/package.json#L91-L93).
 */
export const ExampleDismissingToasts: Story = {
  render: (args) => <ToastNotificationManager {...args} />,
  parameters: {
    // For interactive use, low value in snap testing again since already covered in other stories.
    chromatic: { disableSnapshot: true },
    snapshot: { skip: true },
  },
};
