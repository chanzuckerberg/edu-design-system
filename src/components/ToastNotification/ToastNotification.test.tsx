import { generateSnapshots } from '@chanzuckerberg/story-utils';
import { composeStories } from '@storybook/react';
import { render, waitFor } from '@testing-library/react';

import React from 'react';
import { ToastNotification } from './ToastNotification';
import * as stories from './ToastNotification.stories';
import type { StoryFile } from '../../util/utility-types';

const { AutoDismiss } = composeStories(stories);

const { AutoDismiss: skip, ...staticStories } = stories;

describe('<ToastNotification />', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  generateSnapshots(staticStories as StoryFile);

  it('triggers the onDissmiss after a delay', async () => {
    const consoleSpy = jest.spyOn(console, 'log').mockImplementation();

    render(<AutoDismiss />);

    await waitFor(() => expect(consoleSpy).toHaveBeenCalledTimes(1));
  });

  describe('emits messages when misused', () => {
    let consoleErrorMock: jest.SpyInstance;
    beforeEach(() => {
      consoleErrorMock = jest.spyOn(console, 'error');
      consoleErrorMock.mockImplementation();
    });

    it('generates an error when onDissmiss and type=auto are misused', async () => {
      // One must use onDismiss if auto is used
      render(
        <ToastNotification dismissType="auto" timeout={50} title="test" />,
      );

      await waitFor(() => expect(consoleErrorMock).toHaveBeenCalledTimes(1));
    });

    it('[EDS-1453] generates an error message when using the dissmissType prop', async () => {
      render(<ToastNotification dissmissType="manual" title="test" />);

      await waitFor(() => expect(consoleErrorMock).toHaveBeenCalledTimes(1));
    });

    it('[EDS-1453] generates no error when using the correct dismissType prop', async () => {
      render(<ToastNotification dismissType="manual" title="test" />);

      await waitFor(() => expect(consoleErrorMock).toHaveBeenCalledTimes(0));
    });
  });
});
