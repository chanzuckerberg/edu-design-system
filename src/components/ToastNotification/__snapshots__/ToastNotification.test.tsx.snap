// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToastNotification /> Critical story renders snapshot 1`] = `
<div
  class="toast toast--status-critical w-[384px]"
>
  <svg
    aria-hidden="true"
    class="icon toast__icon"
    fill="currentColor"
    height="1.875rem"
    style="--icon-size: 1.875rem;"
    viewBox="0 0 24 24"
    width="1.875rem"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M14.9,3H9.1C8.57,3,8.06,3.21,7.68,3.59l-4.1,4.1C3.21,8.06,3,8.57,3,9.1v5.8c0,0.53,0.21,1.04,0.59,1.41l4.1,4.1 C8.06,20.79,8.57,21,9.1,21h5.8c0.53,0,1.04-0.21,1.41-0.59l4.1-4.1C20.79,15.94,21,15.43,21,14.9V9.1c0-0.53-0.21-1.04-0.59-1.41 l-4.1-4.1C15.94,3.21,15.43,3,14.9,3z M15.54,15.54L15.54,15.54c-0.39,0.39-1.02,0.39-1.41,0L12,13.41l-2.12,2.12 c-0.39,0.39-1.02,0.39-1.41,0l0,0c-0.39-0.39-0.39-1.02,0-1.41L10.59,12L8.46,9.88c-0.39-0.39-0.39-1.02,0-1.41l0,0 c0.39-0.39,1.02-0.39,1.41,0L12,10.59l2.12-2.12c0.39-0.39,1.02-0.39,1.41,0l0,0c0.39,0.39,0.39,1.02,0,1.41L13.41,12l2.12,2.12 C15.93,14.51,15.93,15.15,15.54,15.54z"
    />
  </svg>
  <div
    class="toast__body"
  >
    <span
      class="text text--title-md toast__text"
    >
      You've got a temporary notification!
    </span>
  </div>
</div>
`;

exports[`<ToastNotification /> Default story renders snapshot 1`] = `
<div
  class="toast toast--status-favorable w-[384px]"
>
  <svg
    aria-hidden="true"
    class="icon toast__icon"
    fill="currentColor"
    height="1.875rem"
    style="--icon-size: 1.875rem;"
    viewBox="0 0 24 24"
    width="1.875rem"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M12 22C10.6167 22 9.31667 21.7373 8.1 21.212C6.88333 20.6873 5.825 19.975 4.925 19.075C4.025 18.175 3.31267 17.1167 2.788 15.9C2.26267 14.6833 2 13.3833 2 12C2 10.6167 2.26267 9.31667 2.788 8.1C3.31267 6.88333 4.025 5.825 4.925 4.925C5.825 4.025 6.88333 3.31233 8.1 2.787C9.31667 2.26233 10.6167 2 12 2C13.3833 2 14.6833 2.26233 15.9 2.787C17.1167 3.31233 18.175 4.025 19.075 4.925C19.975 5.825 20.6873 6.88333 21.212 8.1C21.7373 9.31667 22 10.6167 22 12C22 13.3833 21.7373 14.6833 21.212 15.9C20.6873 17.1167 19.975 18.175 19.075 19.075C18.175 19.975 17.1167 20.6873 15.9 21.212C14.6833 21.7373 13.3833 22 12 22ZM10.6 16.6L17.65 9.55L16.25 8.15L10.6 13.8L7.75 10.95L6.35 12.35L10.6 16.6Z"
    />
  </svg>
  <div
    class="toast__body"
  >
    <span
      class="text text--title-md toast__text"
    >
      You've got a temporary notification!
    </span>
  </div>
</div>
`;

exports[`<ToastNotification /> Favorable story renders snapshot 1`] = `
<div
  class="toast toast--status-favorable w-[384px]"
>
  <svg
    aria-hidden="true"
    class="icon toast__icon"
    fill="currentColor"
    height="1.875rem"
    style="--icon-size: 1.875rem;"
    viewBox="0 0 24 24"
    width="1.875rem"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M12 22C10.6167 22 9.31667 21.7373 8.1 21.212C6.88333 20.6873 5.825 19.975 4.925 19.075C4.025 18.175 3.31267 17.1167 2.788 15.9C2.26267 14.6833 2 13.3833 2 12C2 10.6167 2.26267 9.31667 2.788 8.1C3.31267 6.88333 4.025 5.825 4.925 4.925C5.825 4.025 6.88333 3.31233 8.1 2.787C9.31667 2.26233 10.6167 2 12 2C13.3833 2 14.6833 2.26233 15.9 2.787C17.1167 3.31233 18.175 4.025 19.075 4.925C19.975 5.825 20.6873 6.88333 21.212 8.1C21.7373 9.31667 22 10.6167 22 12C22 13.3833 21.7373 14.6833 21.212 15.9C20.6873 17.1167 19.975 18.175 19.075 19.075C18.175 19.975 17.1167 20.6873 15.9 21.212C14.6833 21.7373 13.3833 22 12 22ZM10.6 16.6L17.65 9.55L16.25 8.15L10.6 13.8L7.75 10.95L6.35 12.35L10.6 16.6Z"
    />
  </svg>
  <div
    class="toast__body"
  >
    <span
      class="text text--title-md toast__text"
    >
      You've got a temporary notification!
    </span>
  </div>
</div>
`;

exports[`<ToastNotification /> NotDismissable story renders snapshot 1`] = `
<div
  class="toast toast--status-favorable w-[384px]"
>
  <svg
    aria-hidden="true"
    class="icon toast__icon"
    fill="currentColor"
    height="1.875rem"
    style="--icon-size: 1.875rem;"
    viewBox="0 0 24 24"
    width="1.875rem"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M12 22C10.6167 22 9.31667 21.7373 8.1 21.212C6.88333 20.6873 5.825 19.975 4.925 19.075C4.025 18.175 3.31267 17.1167 2.788 15.9C2.26267 14.6833 2 13.3833 2 12C2 10.6167 2.26267 9.31667 2.788 8.1C3.31267 6.88333 4.025 5.825 4.925 4.925C5.825 4.025 6.88333 3.31233 8.1 2.787C9.31667 2.26233 10.6167 2 12 2C13.3833 2 14.6833 2.26233 15.9 2.787C17.1167 3.31233 18.175 4.025 19.075 4.925C19.975 5.825 20.6873 6.88333 21.212 8.1C21.7373 9.31667 22 10.6167 22 12C22 13.3833 21.7373 14.6833 21.212 15.9C20.6873 17.1167 19.975 18.175 19.075 19.075C18.175 19.975 17.1167 20.6873 15.9 21.212C14.6833 21.7373 13.3833 22 12 22ZM10.6 16.6L17.65 9.55L16.25 8.15L10.6 13.8L7.75 10.95L6.35 12.35L10.6 16.6Z"
    />
  </svg>
  <div
    class="toast__body"
  >
    <span
      class="text text--title-md toast__text"
    >
      You've got a temporary notification!
    </span>
  </div>
</div>
`;
